src/programs/bamfillquery.1
.TH BAMFILLQUERY 1 "August 2017" BIOBAMBAM
.SH NAME
bamfillquery - fill query sequences into BAM files
.SH SYNOPSIS
.PP
.B bamfillquery
[options] <in.bam queries.fasta >out.bam
.SH DESCRIPTION
bamfillquery reads a SAM/BAM/CRAM file and a FastA file, copies the
sequences found in the FastA file into the query sequence field of the
SAM/BAM/CRAM file and writes the resulting data
to a SAM/BAM/CRAM file. Both the alignment input file
and the FastA file need to be sorted by query name. See bamsort or
bamsormadup for sorting the SAM/BAM/CRAM file to this order.
.PP
The following key=value pairs can be given:
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B verbose=<1>:
Valid values are
.IP 1:
print progress report on standard error
.IP 0:
do not print progress report
.PP
.B tmpfile=<filename>: 
prefix for temporary files. By default the temporary files are created in the current directory
.PP
.B md5=<0|1>:
md5 checksum creation for output file. Valid values are
.IP 0:
do not compute checksum. This is the default.
.IP 1:
compute checksum. If the md5filename key is set, then the checksum is
written to the given file. If md5filename is unset, then no checksum will be computed.
.PP
.B md5filename
file name for md5 checksum if md5=1.
.PP
.B index=<0|1>:
compute BAM index for output file. Valid values are
.IP 0:
do not compute BAM index. This is the default.
.IP 1:
compute BAM index. If the indexfilename key is set, then the BAM index is
written to the given file. If indexfilename is unset, then no BAM index will be computed.
.PP
.B indexfilename
file name for BAM index if index=1.
.PP
.B inputformat=<bam>: input file format.
All versions of bamsort come with support for the BAM input format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit)
.PP
.B outputformat=<bam>: output file format.
All versions of bamsort come with support for the BAM output format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit). This format is not advisable for data sorted by query name.
.PP
.B I=<[stdin]>: 
input filename, standard input if unset.
.PP
.B O=<[stdout]>: 
output filename, standard output if unset.
.PP
.B inputthreads=<[1]>:
input helper threads, only valid for inputformat=bam.
.PP
.B outputthreads=<[1]>:
output helper threads, only valid for outputformat=bam.
.PP
.B reference=<[]>:
reference FastA file for inputformat=cram and outputformat=cram. An index file (.fai) is required. 
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2016 German Tischler, \(co 2011-2013 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.76) unstable; urgency=medium

  * add rcsupport switch in bamsormadup

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 08 Aug 2017 16:49:12 +0200

biobambam2 (2.0.75) unstable; urgency=medium

  * add mc field in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 03 Aug 2017 12:24:54 +0200

biobambam2 (2.0.74) unstable; urgency=medium

  * add bamfillquery

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 Aug 2017 10:00:47 +0200

biobambam2 (2.0.73) unstable; urgency=medium

  * add bamfilterrefid
  * allow O key in bammerge (disable check for stdout being a terminal if O key is set)
  * add bamrefinterval program

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 20 Jun 2017 14:48:51 +0200

biobambam2 (2.0.72) unstable; urgency=medium

  * add output file prefix for default read group in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:22:23 +0100

biobambam2 (2.0.71) unstable; urgency=medium

  * add outputperreadgrouprgsm and outputperreadgroupprefix options in bamtofastq

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Mar 2017 10:06:40 +0100

biobambam2 (2.0.70) unstable; urgency=medium

  * add bamscrapcount

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 07 Mar 2017 17:05:10 +0100

biobambam2 (2.0.69) unstable; urgency=medium

  * encode ref id of first read in file in bamexplode file names

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 13:32:37 +0100

biobambam2 (2.0.68) unstable; urgency=medium

  * added HD line header merging in bamcat

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Jan 2017 11:33:22 +0100

biobambam2 (2.0.67) unstable; urgency=medium

  * no adapter clipping if remaining sequence would be a single base (fix by Andrew Whitwham)
  * add bamauxmerge

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Jan 2017 17:30:44 +0100

biobambam2 (2.0.66) unstable; urgency=medium

  * add parserbuffersize parameter in bamsormadup 
  * add optminpixeldif parameter passing in bamsormadup
  * add bamfilterlength program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 05 Jan 2017 12:06:53 +0100

biobambam2 (2.0.65) unstable; urgency=medium

  * fix bug in bammarkduplicatesopt where last set of duplicates was no marked with od fields

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 30 Nov 2016 12:14:14 +0100

biobambam2 (2.0.64) unstable; urgency=medium

  * add addmatecigar option in bammarkduplicatesopt

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 29 Nov 2016 12:31:07 +0100

biobambam2 (2.0.63) unstable; urgency=medium

  * fix bug in bammarkduplicates/bammarkduplicates2 concerning input via pipe, markthreads>1 and rewritebam=2 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 28 Nov 2016 10:47:09 +0100

biobambam2 (2.0.62) unstable; urgency=medium

  * fix bin computation for unmapped reads in bamsormadup with SAM input 

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 Nov 2016 13:34:04 +0100

biobambam2 (2.0.61) unstable; urgency=medium

  * filter out broken alignments (cigar reaching past end of reference) in bamrecalculatecigar instead of quitting

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 21 Nov 2016 09:06:09 +0100

biobambam2 (2.0.60) unstable; urgency=medium

  * add bamnumericalindex program

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 17 Nov 2016 15:13:15 +0100

biobambam2 (2.0.59) unstable; urgency=medium

  * adapt bamparsort to new libmaus2 API
  * add numeric index construction in bamindex

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 14 Nov 2016 14:09:29 +0100

biobambam2 (2.0.58) unstable; urgency=medium

  * Update to libmaus2 version 2.0.268 (fixes for mate cigar string copying in bamsormadup)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 01 Nov 2016 10:35:57 +0100

biobambam2 (2.0.57) unstable; urgency=medium

  * Added an explicit iRODS disconnect to avoid the iRODS 4.19 crash. (patch by Andrew Whitwham)

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 18 Sep 2016 18:56:06 +0200

biobambam2 (2.0.56) unstable; urgency=medium

  * fix compilation after API change in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 25 Aug 2016 15:51:00 +0200

biobambam2 (2.0.55) unstable; urgency=medium

  * conditional compilation for PTHREAD_MUTEX_RECURSIVE/PTHREAD_MUTEX_RECURSIVE_NP

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 03 Aug 2016 12:12:34 +0200

biobambam2 (2.0.54) unstable; urgency=medium

  * implement the tags option in bamtofastq (support copying of aux fields from SAM/BAM/CRAM to FastQ)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 02 Aug 2016 16:25:48 +0200

biobambam2 (2.0.53) unstable; urgency=medium

  * update to libmaus2 version 2.0.240
  * merge/import Andrew Whitwham's parallel adapter finding code

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 01 Aug 2016 10:41:37 +0200

biobambam2 (2.0.52) unstable; urgency=medium

  * add bamfiltermc program

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 12:39:39 +0200

biobambam2 (2.0.51) unstable; urgency=medium

  * only copy mate cigar string if mate is mapped in bamsort

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 15 Jul 2016 10:43:26 +0200

biobambam2 (2.0.50) unstable; urgency=medium

  * do not copy CIGAR data of unmapped mates in bamsormadup 

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jul 2016 16:13:07 +0200

biobambam2 (2.0.49) unstable; urgency=medium

  * update to libmaus2 version 2.0.216 

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 29 Jun 2016 10:38:45 +0200

biobambam2 (2.0.48) unstable; urgency=medium

  * use AdapterFilter object as const in bamadapterfind
  * add blocksortverbose parameter in bamsormadup
  * reflect API update in libmaus2

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 28 Jun 2016 12:28:37 +0200

biobambam2 (2.0.47) unstable; urgency=medium

  * filter out alignments extending past the end of a reference sequence in filtersam

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Jun 2016 12:22:09 +0200


src/programs/bamfillquery.cpp
/**
    bambam
    Copyright (C) 2009-2017 German Tischler
    Copyright (C) 2011-2013 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include "config.h"

#include <iostream>
#include <queue>

#include <libmaus2/bambam/BamPeeker.hpp>
#include <libmaus2/fastx/FastaPeeker.hpp>
#include <libmaus2/bambam/BamAlignment.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BamDecoder.hpp>
#include <libmaus2/bambam/BamWriter.hpp>
#include <libmaus2/bambam/ProgramHeaderLineSet.hpp>
#include <libmaus2/bambam/BamMultiAlignmentDecoderFactory.hpp>
#include <libmaus2/bambam/StrCmpNum.hpp>

#include <libmaus2/util/ArgInfo.hpp>

#include <biobambam2/Licensing.hpp>
#include <biobambam2/AttachRank.hpp>

static int getDefaultLevel() { return Z_DEFAULT_COMPRESSION; }
static int getDefaultVerbose() { return 1; }
static std::string getDefaultInputFormat() { return "bam"; }

#include <libmaus2/lz/BgzfDeflateOutputCallbackMD5.hpp>
#include <libmaus2/bambam/BgzfDeflateOutputCallbackBamIndex.hpp>
static int getDefaultMD5() { return 0; }
static int getDefaultIndex() { return 0; }

int bamfillquery(::libmaus2::util::ArgInfo const & arginfo)
{
	if ( arginfo.getNumRestArgs() < 1 )
	{
		::libmaus2::exception::LibMausException se;
		se.getStream() << "[E] required first argument (FastA containing query sequences) is missing" << std::endl;
		se.finish();
		throw se;
	}

	// int const level = libmaus2::bambam::BamBlockWriterBaseFactory::checkCompressionLevel(arginfo.getValue<int>("level",getDefaultLevel()));
	int const verbose = arginfo.getValue<int>("verbose",getDefaultVerbose());

	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type decwrapper(
		libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(
			arginfo,
			false, // do not put rank
			0, /* copy stream */
			std::cin /* standard input */
		)
	);
	::libmaus2::bambam::BamAlignmentDecoder & dec = decwrapper->getDecoder();

	libmaus2::bambam::BamPeeker BP(dec);
	::libmaus2::bambam::BamHeader const & header = BP.dec.getHeader();
	libmaus2::fastx::FastaPeeker FP(arginfo.getUnparsedRestArg(0));

	std::string const headertext(header.text);

	// add PG line to header
	std::string const upheadtext = ::libmaus2::bambam::ProgramHeaderLineSet::addProgramLine(
		headertext,
		"bamfillquery", // ID
		"bamfillquery", // PN
		arginfo.commandline, // CL
		::libmaus2::bambam::ProgramHeaderLineSet(headertext).getLastIdInChain(), // PP
		std::string(PACKAGE_VERSION) // VN
	);

	// construct new header
	libmaus2::bambam::BamHeader const uphead(upheadtext);

	/*
	 * start index/md5 callbacks
	 */
	std::string const tmpfilenamebase = arginfo.getValue<std::string>("tmpfile",arginfo.getDefaultTmpFileName());
	std::string const tmpfileindex = tmpfilenamebase + "_index";
	::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);

	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo) != std::string() )
			md5filename = libmaus2::bambam::BamBlockWriterBaseFactory::getMD5FileName(arginfo);
		else
			std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

		if ( md5filename.size() )
		{
			::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
			Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
			cbs.push_back(Pmd5cb.get());
		}
	}
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
	{
		if ( libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo) != std::string() )
			indexfilename = libmaus2::bambam::BamBlockWriterBaseFactory::getIndexFileName(arginfo);
		else
			std::cerr << "[V] no filename for index given, not creating index" << std::endl;

		if ( indexfilename.size() )
		{
			libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
			Pindex = UNIQUE_PTR_MOVE(Tindex);
			cbs.push_back(Pindex.get());
		}
	}
	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
	if ( cbs.size() )
		Pcbs = &cbs;
	/*
	 * end md5/index callbacks
	 */

	libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Uout ( libmaus2::bambam::BamBlockWriterBaseFactory::construct(uphead, arginfo, Pcbs) );
	libmaus2::bambam::BamBlockWriterBase & Pout = *Uout;

	libmaus2::bambam::BamAlignment algn;
	uint64_t c = 0;

	libmaus2::fastx::FastAReader::pattern_type pat;
	std::string prevname;
	bool prevnamevalid = false;
	std::string prevbamname;
	bool prevbamnamevalid = false;

	while ( FP.peekNext(pat) )
	{
		FP.getNext(pat);
		std::string const name = pat.getShortStringId();

		if ( prevnamevalid )
		{
			int const r = libmaus2::bambam::StrCmpNum::strcmpnum(prevname.c_str(),name.c_str());

			if ( r >= 0 )
			{
				::libmaus2::exception::LibMausException se;
				se.getStream() << "[E] FastA file is not sorted by query name" << std::endl;
				se.getStream() << "[E] " << prevname << std::endl;
				se.getStream() << "[E] " << name << std::endl;
				se.finish();
				throw se;
			}
		}

		if ( verbose )
			std::cerr << name << std::endl;

		// check order in alignment file
		if ( prevbamnamevalid )
		{
			if ( BP.peekNext(algn) )
			{
				if ( libmaus2::bambam::StrCmpNum::strcmpnum(algn.getName(),prevbamname.c_str()) < 0 )
				{
					::libmaus2::exception::LibMausException se;
					se.getStream() << "[E] Alignment stream is not sorted by query name" << std::endl;
					se.getStream() << "[E] " << algn.getName() << std::endl;
					se.finish();
					throw se;
				}
			}
		}

		if ( BP.peekNext(algn) && libmaus2::bambam::StrCmpNum::strcmpnum(algn.getName(),name.c_str()) < 0 )
		{
			::libmaus2::exception::LibMausException se;
			se.getStream() << "[E] Alignment stream contains query name not cotained in FastA input or files are not properly sorted" << std::endl;
			se.getStream() << "[E] " << algn.getName() << std::endl;
			se.finish();
			throw se;
		}

		while ( BP.peekNext(algn) && algn.getName() == name )
		{
			BP.getNext(algn);

			std::string seq;
			if ( algn.isReverse() )
			{
				seq = libmaus2::fastx::reverseComplementUnmapped(libmaus2::fastx::remapString(libmaus2::fastx::mapString(pat.spattern)));
			}
			else
			{
				seq = libmaus2::fastx::remapString(libmaus2::fastx::mapString(pat.spattern));
			}

			if ( algn.isMapped() )
			{
				uint64_t const fc = algn.getFrontHardClipping();
				uint64_t const bc = algn.getBackHardClipping();
				uint64_t const l = seq.size() - (fc+bc);

				assert ( fc + bc <= seq.size() );

				seq = seq.substr(fc,l);
			}

			algn.replaceSequence(
				seq,
				std::string(seq.size(),'H')
			);

			Pout.writeAlignment(algn);

			prevbamnamevalid = true;
			prevbamname = algn.getName();

			++c;
		}

		prevname = name;
		prevnamevalid = true;
	}

	Uout.reset();

	if ( Pmd5cb )
	{
		Pmd5cb->saveDigestAsFile(md5filename);
	}
	if ( Pindex )
	{
		Pindex->flush(std::string(indexfilename));
	}

	return EXIT_SUCCESS;
}

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( "verbose=<["+::biobambam2::Licensing::formatNumber(getDefaultVerbose())+"]>", "print progress information" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5=<["+::biobambam2::Licensing::formatNumber(getDefaultMD5())+"]>", "create md5 check sum (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5filename=<filename>", "file name for md5 check sum (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "index=<["+::biobambam2::Licensing::formatNumber(getDefaultIndex())+"]>", "create BAM index (default: 0)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "indexfilename=<filename>", "file name for BAM index file (default: extend output file name)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "tmpfile=<filename>", "prefix for temporary files, default: create files in current directory" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", std::string("input format (") + libmaus2::bambam::BamMultiAlignmentDecoderFactory::getValidInputFormats() + ")" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("outputformat=<[")+libmaus2::bambam::BamBlockWriterBaseFactory::getDefaultOutputFormat()+"]>", std::string("output format (") + libmaus2::bambam::BamBlockWriterBaseFactory::getValidOutputFormats() + ")" ) );
				V.push_back ( std::pair<std::string,std::string> ( "I=<[stdin]>", "input filename (standard input if unset)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "O=<[stdout]>", "output filename (standard output if unset)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "inputthreads=<[1]>", "input helper threads (for inputformat=bam only, default: 1)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "outputthreads=<[1]>", "output helper threads (for outputformat=bam only, default: 1)" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;

				std::cerr << "The keep and remove keys are mutually exclusive. Tags are given by their two character ids. Multiple ids are separated by commas." << std::endl;

				return EXIT_SUCCESS;
			}

		return bamfillquery(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
